monkey/monkey_island/cc/ui/src/components/contexts/plugins/PluginsContext.tsx

```diff
@@ -1,107 +1,173 @@
 import React, {createContext, useEffect, useState} from 'react';
 import AuthComponent from '../../AuthComponent';
 import semver from 'semver';
-import _ from 'lodash';
 
-export const PluginsContext = createContext({});
+
+// Types returned from the API
+type PluginMetadata = {
+  name: string,
+  type_: string,
+  resource_path: string,
+  sha256: string,
+  description: string,
+  version: string,
+  safe: boolean
+}
+
+type PluginMetadataResponse = {
+  [key: string]: {[key: string]: PluginMetadata[]}
+}
+
+type PluginFromManifest = {
+  name: string,
+  plugin_type: string,
+  supported_operating_systems: string[],
+  target_operating_systems: string[],
+  title: string,
+  version: string,
+  description: string,
+  remediation_suggestion?: string,
+  link_to_documentation?: string,
+  safe: boolean
+}
+
+type PluginManifestResponse = {
+  [key: string]: {[key: string]: PluginFromManifest[]}
+}
+
+// Types used in the UI
+export type AgentPlugin = {
+  id: string,
+  name: string,
+  pluginType: string,
+  description: string,
+  safe: boolean,
+  version: string,
+}
+
+export type InstalledPlugin = AgentPlugin & {
+  title: string,
+  supportedOperatingSystems: string[],
+  targetOperatingSystems: string[],
+  linkToDocumentation?: string,
+  remediationSuggestion?: string,
+}
+
+export type AvailablePlugin = AgentPlugin & {
+  resourcePath: string,
+  sha256: string,
+}
+
+export type PluginsContextType = {
+  availablePlugins: AvailablePlugin[],
+  installedPlugins: InstalledPlugin[],
+  numberOfPluginsThatRequiresUpdate: number,
+  refreshAvailablePlugins: () => Promise<void>,
+  refreshInstalledPlugins: () => Promise<void>,
+  refreshNumberOfUpgradablePlugins: () => Promise<void>,
+  setInstalledPlugins: (installedPlugins: InstalledPlugin[]) => void,
+  setAvailablePlugins: (availablePlugins: AvailablePlugin[]) => void,
+}
+
+export const PluginsContext :null | PluginsContextType  = createContext(null);
+
+export const generatePluginId = (name, type, version) => {
+  return `${name}${type}${version}`;
+}
 
 const authComponent = new AuthComponent({});
 
-export const PluginState = () => {
-  const [allPlugins, setAllPlugins] = useState([]);
-  const [allInstalledPlugins, setAllInstalledPlugins] = useState([]);
```

Comment:
Not needed, installed plugins and all downloadable plugins are the only states we need